httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ForwardedRequest.java

```diff
@@ -0,0 +1,189 @@
+     * @throws IOException       if an IO error occurs while processing the request
+     * @throws ProtocolException if the request authority is not specified
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context) throws HttpException, IOException {
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+
+        final ProtocolVersion ver = context.getProtocolVersion() != null ? context.getProtocolVersion() : HttpVersion.HTTP_1_1;
+
+        final URIAuthority authority = request.getAuthority();
+        if (authority == null) {
+            throw new ProtocolException("Request authority not specified");
+        }
+
+        final int port = authority.getPort();
+
+        final StringBuilder valueBuilder = new StringBuilder();
+        final String forwardedHeaderValue = request.getFirstHeader(FORWARDED_HEADER_NAME) != null ?
+                request.getFirstHeader(FORWARDED_HEADER_NAME).getValue() : null;
+
+        if (forwardedHeaderValue != null) {
+            // Forwarded header already exists, add current hop details to the end of the existing value
+            valueBuilder.append(forwardedHeaderValue);
+            valueBuilder.append(", ");
+        }
+
+        // Add the current hop details
+        final EndpointDetails endpointDetails = (EndpointDetails) context.getAttribute(HttpCoreContext.CONNECTION_ENDPOINT);
+
+        // Add the "by" parameter
+        if (endpointDetails != null) {
+            final SocketAddress remoteAddress = endpointDetails.getRemoteAddress();
+            if (remoteAddress instanceof InetSocketAddress) {
+                final InetSocketAddress inetAddress = (InetSocketAddress) remoteAddress;
+                final String byValue = inetAddress.getHostString() + ":" + inetAddress.getPort();
+                if (inetAddress.getAddress() instanceof Inet6Address) {
+                    valueBuilder.append("by=\"").append(byValue).append("\"");
+                } else {
+                    valueBuilder.append("by=").append(byValue);
+                }
+            }
+        }
+
+        // Add the "for" parameter
+        if (endpointDetails != null) {
+            final SocketAddress localAddress = endpointDetails.getLocalAddress();
+            if (localAddress instanceof InetSocketAddress) {
+                final InetSocketAddress inetAddress = (InetSocketAddress) localAddress;
+                final String forValue = inetAddress.getHostString() + ":" + inetAddress.getPort();
+                if (inetAddress.getAddress() instanceof Inet6Address) {
+                    valueBuilder.append(";for=\"").append(forValue).append("\"");
+                } else {
+                    valueBuilder.append(";for=").append(forValue);
+                }
+            }
+        }
+
+        // Add the "host" parameter
+        final String hostValue = authority.getHostName();
```

Comment:
@arturobernalg You can safely assume `#getHostName` of the authority is never `null`.